library/psa_crypto.c

```diff
@@ -2084,31 +2085,49 @@ psa_status_t psa_copy_key( mbedtls_svc_key_id_t source_key,
                                      &target_slot, &driver );
     if( status != PSA_SUCCESS )
         goto exit;
-
-#if defined(MBEDTLS_PSA_CRYPTO_SE_C)
-    if( driver != NULL )
+    if( PSA_KEY_LIFETIME_GET_LOCATION( target_slot->attr.lifetime ) !=
+        PSA_KEY_LIFETIME_GET_LOCATION( source_slot->attr.lifetime ) )
     {
-        /* Copying to a secure element is not implemented yet. */
-        status = PSA_ERROR_NOT_SUPPORTED;
+        /*
+         * If the source and target keys are stored across different locations,
+         * the source key would need to be exported as plaintext and re-imported
+         * in the other location. This has security implications which have not
+         * been fully mapped.For now, this can be acheived through
+         * appropriate API invocations from the application, if needed.
+         * */
+        status = PSA_ERROR_INVALID_ARGUMENT;
         goto exit;
     }
-#endif /* MBEDTLS_PSA_CRYPTO_SE_C */
-
+    /*
+     * When the source and target keys are within the same location,
+     * - For transparent keys it is a blind copy sans any driver invocation,
+     * - For opaque keys this translates to an invocation of the drivers'
+     *   copy_key entry point through the dispatch layer.
+     * */
     if( psa_key_lifetime_is_external( actual_attributes.core.lifetime ) )
```

Comment:
Why only dispatch to the driver wrapper for an external-to-same copy? Other key management functions dispatch through the driver wrapper for transparent keys as well.

Reply:
Well for transparent keys, there is no need for the driver to be involved at all, as long as the lifetime is the same. That's the reason the driver dispatch layer itself is avoided here.
Do we still want to hit the driver dispatch layer, not look for transparent drivers and do an internal implementation ?
I thought that was not needed.

Reply:
I thought the architecture should be uniform for all key management functions. But on second thoughts, copy is fundamentally different since there's no involvement of a driver at all for transparent keys. The dispatch layer would not call anything other than the built-in implementation for copy. So good choice, indeed, the copy dispatch is not like the others.